### PR TITLE
Upgrade all the CI runners to Ubuntu 22.04

### PR DESCRIPTION
**Pull request by dbussink**
This sets things explicitly to Ubuntu 22.04, which is the latest LTS release. We also set this for things on ubuntu-latest right now, so we don't see automatic upgrades but at our own pace and we know when we have to fix things / don't have to rush it.

We have still plenty of time with 20.04, but rather do it now as GitHub has made 22.04 the stable for ubuntu-latest already.

The upgrade should be a lot easier than 18.04 to 20.04 since that also included a change of the default MySQL version from 5.7 to 8.0. With 22.04 we're still on 8.0 and the same version there, so not as much changes at all.

With upgrading early, we also don't have to backport it this time and rush it into older releases as well. Since all our supported releases will be EOL before Ubuntu 20.04 will be EOL, so we can just let it stay on 20.04. This is another big advantage of upgrading earlier rather than later.

Also updated to the latest `mysql-apt-config` packages everywhere (which were very inconsistent versioning wise anyway).

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by dbussink** [comment]
@deepthi @mattlord I'm wondering here, the reason this is failing CI for mariadb is that Ubuntu 22.04 has upgraded the default MariaDB from 10.3 to 10.6. We don't support 10.6, nor do we plan to, so should we remove the MariaDB 10.3 unit test job here? 

We still want to have it working for external tablets and movetables, but I don't think that's something we'd want / need the unit tests for but instead would need a specific test for that?

**Comment by mattlord** [comment]
> @deepthi @mattlord I'm wondering here, the reason this is failing CI for mariadb is that Ubuntu 22.04 has upgraded the default MariaDB from 10.3 to 10.6.

Yeah, that would break things as the newer MariaDB init process is not MySQL/Vitess compatible. We'd then have to install 10.3 ourselves rather than use the system packages.

> We don't support 10.6, nor do we plan to, so should we remove the MariaDB 10.3 unit test job here?
> 
> We still want to have it working for external tablets and movetables, but I don't think that's something we'd want / need the unit tests for but instead would need a specific test for that?

I definitely wanted to delete it at some point. If we do, then I'll prioritize this and work on it ASAP: https://github.com/vitessio/vitess/issues/10732

Getting rid of the MariaDB unit tests is probably overdue at this point so I'm good getting rid of it here and starting on that work. Just let me know.

**Comment by dbussink** [comment]
> I definitely wanted to delete it at some point. If we do, then I'll prioritize this and work on it ASAP: #10732
> 
> Getting rid of the MariaDB unit tests is probably overdue at this point so I'm good getting rid of it here and starting on that work. Just let me know.

I think this is great. Once we land #10732, I can update this PR so we do this upgrade as well with no other issues :smile:.

**Comment by dbussink** [comment]
@mattlord Hmm, I guess the vreplication test here now fails since it can't start / manage MariaDB since the tests do use our own tools to manage that process then? Like we're using `mysqlctl` here to start it? 

```
I0105 10:48:41.351921   18941 cluster_test.go:506] Waiting for mysql process for tablet zone1-100
[88](https://github.com/vitessio/vitess/actions/runs/3845936691/jobs/6550666729#step:12:89)
    cluster_test.go:508: exit status 1 :: Unable to start mysql server for &{zone1-100 vttablet /tmp/vt_2314915630/vreple2e_545593/tmp/vt_0000000100_querylog.txt 100 zone1-0000000100  16100 17091 0 {vtctl vtctl  etcd2 127.0.0.1:2379 /vitess/global 127.0.0.1:2379 / 16} /tmp/vt_2314915630/vreple2e_545593/tmp 127.0.0.1 product replica 5 file /tmp/vt_2314915630/vreple2e_545593/backups grpc-queryservice,grpc-tabletmanager,grpc-updatestream,grpc-throttler http://127.0.0.1:15000/ /tmp/vt_2314915630/vreple2e_545593/vt_0000000100 http://127.0.0.1:16100/debug/vars http://127.0.0.1:16100/queryz http://127.0.0.1:16100/debug/status_details false false NOT_SERVING  0 PRIMARY  utf8mb4 http://127.0.0.1:16100/debug/consolidations [--queryserver-config-schema-reload-time 5 --enable-lag-throttler --heartbeat_enable --heartbeat_interval 250ms] <nil> <nil>}
```

**Comment by mattlord** [comment]
> @mattlord Hmm, I guess the vreplication test here now fails since it can't start / manage MariaDB since the tests do use our own tools to manage that process then? Like we're using `mysqlctl` here to start it?
> 
> ```
> I0105 10:48:41.351921   18941 cluster_test.go:506] Waiting for mysql process for tablet zone1-100
> [88](https://github.com/vitessio/vitess/actions/runs/3845936691/jobs/6550666729#step:12:89)
>     cluster_test.go:508: exit status 1 :: Unable to start mysql server for &{zone1-100 vttablet /tmp/vt_2314915630/vreple2e_545593/tmp/vt_0000000100_querylog.txt 100 zone1-0000000100  16100 17091 0 {vtctl vtctl  etcd2 127.0.0.1:2379 /vitess/global 127.0.0.1:2379 / 16} /tmp/vt_2314915630/vreple2e_545593/tmp 127.0.0.1 product replica 5 file /tmp/vt_2314915630/vreple2e_545593/backups grpc-queryservice,grpc-tabletmanager,grpc-updatestream,grpc-throttler http://127.0.0.1:15000/ /tmp/vt_2314915630/vreple2e_545593/vt_0000000100 http://127.0.0.1:16100/debug/vars http://127.0.0.1:16100/queryz http://127.0.0.1:16100/debug/status_details false false NOT_SERVING  0 PRIMARY  utf8mb4 http://127.0.0.1:16100/debug/consolidations [--queryserver-config-schema-reload-time 5 --enable-lag-throttler --heartbeat_enable --heartbeat_interval 250ms] <nil> <nil>}
> ```

What's different about these tests is that they run mysqld in non-standard locations (`/tmp/`):
```
I0105 11:36:47.863458   18165 mysqld.go:631] execCmd: /usr/sbin/mysqld output: /usr/sbin/mysqld  Ver 8.0.31 for Linux on x86_64 (MySQL Community Server - GPL)
[50](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:51)
VT_MYSQL_ROOT is /tmp/mysql-5.7
[51](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:52)
I0105 11:37:32.748310   18165 mysqld.go:617] execCmd: /tmp/mysql-5.7/bin/mysqld /tmp/mysql-5.7/bin/mysqld [--version]
[52](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:53)
I0105 11:37:32.763811   18165 mysqld.go:631] execCmd: /tmp/mysql-5.7/bin/mysqld output: /tmp/mysql-5.7/bin/mysqld  Ver 5.7.37 for linux-glibc2.12 on x86_64 (MySQL Community Server (GPL))
[53](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:54)
MYSQL_FLAVOR is mysql57
[54](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:55)
I0105 11:37:32.763887   18165 cluster_test.go:451] Addshards got 1 shards with [0]
[55](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:56)
I0105 11:37:32.763896   18165 cluster_test.go:461] Adding Shard 0
[56](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:57)
I0105 11:37:32.764307   18165 vtctlclient_process.go:214] Executing vtctlclient with command: vtctlclient --server 127.0.0.1:15999 CreateShard product/0 (attempt 1 of 10)
[57](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:58)
I0105 11:37:32.776607   18165 cluster_test.go:472] Adding Primary tablet
[58](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:59)
I0105 11:37:32.819096   18165 mysqlctl_process.go:149] Starting mysqlctl with command: [mysqlctl --log_dir /tmp/vt_299891415/vreple2e_824043/tmp --tablet_uid 100 --mysql_port 16406 init -- --init_db_sql_file /home/runner/work/vitess/vitess/config/init_db.sql start]
[59](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:60)
I0105 11:37:32.819393   18165 cluster_test.go:506] Waiting for mysql process for tablet zone1-100
[60](https://github.com/vitessio/vitess/actions/runs/3845936650/jobs/6551469384#step:12:61)
    cluster_test.go:508: exit status 1 :: Unable to start mysql server for &{zone1-100 vttablet
```
mysqld fails to start, and given all of this I would be that SELinux / AppArmor is preventing it. We probably need to do something like this in the workflow:
```
sudo ln -s /etc/apparmor.d/usr.sbin.mysqld /etc/apparmor.d/disable/
sudo apparmor_parser -R /etc/apparmor.d/usr.sbin.mysqld
```

Or we could disable AppArmor entirely.

**Comment by dbussink** [comment]
Running things locally it looks like we're missing some libraries:

```
E0105 14:58:03.742165  278677 mysqlctl.go:275] failed init mysql: can't run init_db_sql_file (/home/dirkjan/code/vitessio/vitess/config/init_db.sql): /tmp/mysql-5.7/bin/mysql: exit status 127, output: /tmp/mysql-5.7/bin/mysql: error while loading shared libraries: libncurses.so.5: cannot open shared object file: No such file or directory
```

```
ldd /tmp/mysql-5.7/bin/mysql
	linux-vdso.so.1 (0x00007ffcd81e4000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fee31a01000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fee319fc000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fee319f7000)
	libncurses.so.5 => not found
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fee317cd000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fee316e4000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fee316c4000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fee3149c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fee31a15000)
	libtinfo.so.5 => not found
```

https://github.com/vitessio/vitess/pull/11985/commits/1ac79728fb50e0d3d6484e1c156e601cbc48c9dc should install those. This is because we install the packages from Bionic since there's no Focal / Jammy builds for 5.7.

**Comment by dbussink** [comment]
@deepthi @mattlord For landing this, we need to change the required state for jobs like `Local example using consul on ubuntu-latest` because the base OS is included in the name (since we use a matrix), but that changed here to be consistent on Ubuntu 22.04 everywhere.

**Comment by mattlord** [comment]
> @deepthi @mattlord For landing this, we need to change the required state for jobs like `Local example using consul on ubuntu-latest` because the base OS is included in the name (since we use a matrix), but that changed here to be consistent on Ubuntu 22.04 everywhere.

I would assume that those tests were using ubuntu-latest for some good reason(s). Can we leave those? I would assume that would be Ubuntu 22.10 today.

File: go/mysql/server_flaky_test.go [hunk 1]
```diff
@@ -562,7 +562,8 @@ func TestServer(t *testing.T) {
 	output, err = runMysqlWithErr(t, params, "error after send")
 	require.Error(t, err)
 	assert.Contains(t, output, "ERROR 2013 (HY000)", "Unexpected output for 'panic'")
-	assert.Contains(t, output, "Lost connection to MySQL server during query", "Unexpected output for 'panic'")
+	// MariaDB might not print the MySQL bit here
+	assert.Regexp(t, `Lost connection to( MySQL)? server during query`, output, "Unexpected output for 'panic': %v", output)
```
Comment (mattlord) [inline go/mysql/server_flaky_test.go:566]: 👍 https://mariadb.com/kb/en/the-community-error-2013-hy000-at-line-1353-lost-connection-to-server-durin/

**Comment by dbussink** [comment]
> I would assume that those tests were using ubuntu-latest for some good reason(s). Can we leave those? I would assume that would be Ubuntu 22.10 today.

I don't think there's really super good reasons, except for it being a default? It's not running on 22.10, it's on 22.04. GitHub Actions doesn't provide any non LTS versions (which I totally understand, too much churn for each release). 

Also `ubuntu-latest` is not updated when a new Ubuntu LTS comes out, but separate by GitHub, see also https://github.blog/changelog/2022-11-09-github-actions-ubuntu-latest-workflows-will-use-ubuntu-22-04/.

This is a gradual process though and not all at once, so we're at the mercy of whenever GitHub does that.

> If you and @deepthi prefer to use Ubuntu 22.04 for everything then I'm also fine with that. Other than that the PR looks good so just waiting to confirm on this part. I can then make the GitHub Actions changes for the repo as needed.

I feel pretty strongly about setting things to an explicit version. That allows us to upgrade at a known time and now when GitHub decides to upgrade when maybe we're in the middle of a release and have to scramble then etc. I've always found that explicit versioning but then actually staying up to date (this second part is also important) is a better experience than a floating version that can change at a not exactly known time.

So hence moving it all to 22.04 explicitly in this PR.

**Review by mattlord — APPROVED** [review]
@dbussink OK, thanks! That makes sense to me. I'll approve then. @deepthi can comment/approve as she may be aware of additional history/context that we're not.

**Comment by deepthi** [comment]
> @dbussink OK, thanks! That makes sense to me. I'll approve then. @deepthi can comment/approve as she may be aware of additional history/context that we're not.

I don't think there was ever a really good reason for `ubuntu-latest` except something like "We'll find out if GH changes things and that breaks our tests.". Should be fine for us to pin a particular version.

**Review by deepthi — APPROVED** [review]

**Pull request merged**